src/Microsoft.Health.Dicom.Core/Features/Common/ParallelEnumerable.cs

```diff
@@ -0,0 +1,127 @@
+
+        await foreach (TResult item in state.Items.Reader.ReadAllAsync(cancellationToken))
+        {
+            yield return item;
+        }
+
+        await producer;
+
+        static async ValueTask ProduceAsync(State<TSource, TResult> state)
+        {
+            ChannelWriter<TResult> writer = state.Items.Writer;
+
+            try
+            {
+                // Note: The order is not deterministic.
+                // Items will be produced in the order that they are returned by the selector.
+                await Parallel.ForEachAsync(
+                    state.Source,
+                    state.Options,
+                    (item, token) => ProduceItemAsync(item, state.Selector, writer, token));
+            }
+            catch (Exception e)
+            {
+                writer.Complete(e);
+                return;
+            }
+
+            writer.Complete();
+        }
+
+        static async ValueTask ProduceItemAsync(
+            TSource item,
+            Func<TSource, CancellationToken, ValueTask<TResult>> selector,
+            ChannelWriter<TResult> writer,
+            CancellationToken cancellationToken)
+        {
+            TResult result = await selector(item, cancellationToken);
+            await writer.WriteAsync(result, cancellationToken);
+        }
+    }
+
+    private sealed class State<TSource, TResult>
+    {
+        public State(
+            IEnumerable<TSource> source,
+            Func<TSource, CancellationToken, ValueTask<TResult>> selector,
+            ParallelEnumerationOptions options,
+            CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(options, nameof(options));
+
+            Options = new ParallelOptions
+            {
+                CancellationToken = cancellationToken,
+                MaxDegreeOfParallelism = options.MaxDegreeOfParallelism,
+                TaskScheduler = options.TaskScheduler,
+            };
+
+            Selector = EnsureArg.IsNotNull(selector, nameof(selector));
+            Source = EnsureArg.IsNotNull(source, nameof(source));
+
+            // TODO: Enable buffering with unbounded parallelism if ChannelWriter supports a factory argument.
+            // We have to modify the channel capacity as we will still resolve the selector before blocking
+            // on the channel write. So if the maximum parallelism is 'p', there will be at most 'p' additional
+            // threads blocked after the channel has reached capacity.
+            if (options.MaxBufferedItems <= 0)
+                throw new ArgumentOutOfRangeException(nameof(options));
+
+            if (options.MaxDegreeOfParallelism == -1)
+                throw new ArgumentException(DicomCoreResource.UnsupportedBuffering, nameof(options));
+
+            if (options.MaxBufferedItems <= options.MaxDegreeOfParallelism)
+                throw new ArgumentException(DicomCoreResource.InvalidItemBuffering, nameof(options));
```

Comment:
Perhaps these checks could be put into the ParallelOptions class instead? That way the ParallelOptions class retains its own rules instead of in another class.

Reply:
I've removed the complications here